pallets/did/src/lib.rs

```diff
@@ -28,91 +30,403 @@ mod tests;
-			<DIDs<T>>::insert(sender.clone(), DidRecord::<T> { sign_key, box_key, doc_ref });
-			// deposit an event that the DID has been created
-			Self::deposit_event(RawEvent::DidCreated(sender));
-			Ok(())
+
+	/// Returns the expected signature length (of bytes) for the given
+	/// verification key type.
+	fn get_expected_signature_size(&self) -> usize {
+		match self {
+			PublicVerificationKey::Ed25519(_) | PublicVerificationKey::Sr25519(_) => 64,
+		}
+	}
+}
+
+impl DIDPublicKey for PublicVerificationKey {
+	fn get_did_key_description(&self) -> &'static str {
+		match self {
+			PublicVerificationKey::Ed25519(_) => "Ed25519VerificationKey2018", /* https://w3c.github.io/did-spec-registries/#ed25519verificationkey2018 */
+			PublicVerificationKey::Sr25519(_) => "Sr25519VerificationKey2020", /* Not yet defined in the DID
+			                                                                     * specification. */
 		}
+	}
+}
+
+/// Enum representing the types of encryption keys a DID can control.
+#[derive(Clone, Decode, Debug, Encode, Eq, Ord, PartialEq, PartialOrd)]
+pub enum PublicEncryptionKey {
+	/// An X25519 public key.
+	X55519([u8; 32]),
+}
+
+impl DIDPublicKey for PublicEncryptionKey {
+	fn get_did_key_description(&self) -> &'static str {
+		"X25519KeyAgreementKey2019" // https://w3c.github.io/did-spec-registries/#x25519keyagreementkey2019
+	}
+}
+
+/// All the errors that can be generated when evaluating a DID operation.
+#[derive(Debug, Eq, PartialEq)]
+pub enum DIDError {
+	StorageError(StorageError),
+	SignatureError(SignatureError),
+}
+
+// Used internally to handle storage errors.
+#[derive(Debug, Eq, PartialEq)]
+pub enum StorageError {
+	/// The DID being created is already present on chain.
+	DIDAlreadyPresent,
+	/// The expected DID cannot be found on chain.
+	DIDNotPresent,
+	/// The given DID does not contain the right key to verify the signature of
+	/// a DID operation.
+	VerificationkeyNotPresent(DIDVerificationKeyType),
+}
+
+// Used internally to handle signature errors.
+#[derive(Debug, Eq, PartialEq)]
+pub enum SignatureError {
+	/// The signature is not in the expected format the verification key
+	/// expects.
+	InvalidSignatureFormat,
+	/// The signature is invalid for the payload and the verification key
+	/// provided.
+	InvalidSignature,
+}
+
+/// A trait describing an operation that requires DID authentication.
+pub trait DIDOperation<DIDIdentifier>: Encode
+where
+	DIDIdentifier: Encode + Decode + Clone + Debug + Eq + PartialEq + EncodeLike,
+{
+	/// Returns the type of the verification key to be used to validate the
+	/// operation.
+	fn get_verification_key_type(&self) -> DIDVerificationKeyType;
+	/// Returns the DID identifier of the subject.
+	fn get_did(&self) -> &DIDIdentifier;
+}
+
+/// A DID creation request. It contains the following values:
+/// - the DID identifier being created (only Substrate addresses are allowed in
+///   this version of the pallet);
+/// - the new authentication key to use;
```

Comment:
I would like the `;` to disappear here, never saw this in Rust comments to be honest.

Reply:
Also, usually we start a new comment line "thought" with an uppercase letter - or I would like to enforce this from now on.

Reply:
See below for my suggestion.